### PR TITLE
[fix] OSS: Eager gradient release - free memory

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
- Release the gradients for the model parameters as soon as every rank computes its update, at this point only the gradients for this rank's shard are useful. This frees 100MB of memory for a small model (rn101) & 2 ranks, no other consequence.
- (unrelated) Adjust the circleci config to use the 4 allocated GPUs, probably a more realistic usecase

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
[This](https://app.circleci.com/pipelines/github/facebookresearch/fairscale/511/workflows/08fa3bfb-6bfc-459e-917d-c390f7218565/jobs/1909) shows the apples to apples comparison (the regression test is tuned to the previous memory use, 4475MB for this configuration). 
The two next commits adjust CI to use 4 ranks, this is unrelated and consequently adjust the regression test, the two items are unrelated.

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -367,3 +369,14 @@ def _broadcast_state_dict(self) -> None:
                 # Discard this tensor/rank, broadcast necessary for syncing
                 logging.debug("Discarding broadcast from rank %s", rank)
                 broadcast_object(empty_buffer, src_rank=rank, group=self.group, dist_device=self._device)
+
+    def _free_other_grads(self) -> None:
```
Comment (min-xu-ai) [inline fairscale/optim/oss.py:373]: nice find. perhaps we can do this while doing reduce with other rank? I.e. we can immediately free the grad after they are used? feel free to defer to a separate change.

**Review by min-xu-ai — APPROVED** [review]
minor suggestion.

File: fairscale/optim/oss.py (hunk 1 above)
Comment (blefaudeux) [reply fairscale/optim/oss.py:373]: in sharded_ddp you mean ? Yes I think that there are tons of optimizations to be done there, for now it's pretty slow, which is to be expected. Here I just wanted to fix it in the general case, when step() is called with the normal DDP
